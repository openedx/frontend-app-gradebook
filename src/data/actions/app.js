import { StrictDict } from 'utils';
import { modalFieldKeys, localFilterKeys } from 'data/constants/app';
import { createActionFactory } from './utils';

export const dataKey = 'app';
const createAction = createActionFactory(dataKey);

const closeModal = createAction('closeModal');
/**
 * setCourseId(courseId)
 * loads courseID into local app state
 * @param {string} courseId - course ID from app context
 */
const setCourseId = createAction('setCourseId');

const filterMenu = StrictDict({
  endTransition: createAction('filterMenu/endTransition'),
  startTransition: createAction('filterMenu/startTransition'),
  toggle: createAction('filterMenu/toggle'),
});

/**
  * setModalStateFromTable({ userEntry, subsection })
  * sets modalState
  */
const setModalStateFromTable = createAction('setModalStateFromTable');

/**
 * setSearchValue(searchValue)
 * sets searchValue in local state
 * @param {string} searchValue - search input text
 */
const setSearchValue = createAction('setSearchValue');

const filterReducer = (payload) => (obj, key) => (
  (payload[key] !== undefined) ? { ...obj, [key]: payload[key] } : obj
);

/**
 * setLocalFilter(filterObject)
 * sets a number of localFilterKey fields on localFilters
 * @param {object} filterObject - an object of filter values
 */
const setLocalFilter = createAction('setLocalFilter', (filterObject) => ({
  payload: Object.keys(localFilterKeys).reduce(filterReducer(filterObject), {}),
}));

/**
 * setModalState(modalState)
 * sets a number of modalFieldKey fields on modalState
 * @param {object} modalState - an object of modal state values
 */
const setModalState = createAction('setModalState', (modalState) => ({
  payload: Object.keys(modalFieldKeys).reduce(filterReducer(modalState), {}),
}));

/**
 * setShowImportSuccessToast(shouldShow)
 * Set whether or not to show the Import Grades success toast
 * @param {bool} sholdShow - should show the toast?
 */
const setShowImportSuccessToast = createAction('setShowImportSuccessToast');

/**
 * setView(viewId)
 * sets the UI to display the tab indcated by the passed view id
 * @param {string} viewId - view id as set in app constants
 */
const setView = createAction('setView');

export default StrictDict({
  closeModal,
  filterMenu,
  setCourseId,
  setLocalFilter,
  setModalState,
  setModalStateFromTable,
  setSearchValue,
  setShowImportSuccessToast,
  setView,
});
