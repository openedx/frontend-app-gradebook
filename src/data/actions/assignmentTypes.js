import { StrictDict } from '@src/utils';
import { createActionFactory } from './utils';

export const dataKey = 'assignmentTypes';
const createAction = createActionFactory(dataKey);

const fetching = {
  error: createAction('fetching/error'),
  started: createAction('fetching/started'),
  /**
   * fetching.received(results)
   * @param {object[]} results - assignmentType fetch results
   */
  received: createAction('fetching/received'),
};
/**
 * gotGradesFrozen(areGradesFrozen)
 * @param {bool} are grades frozen?
 */
const gotGradesFrozen = createAction('gotGradesFrozen');

export default StrictDict({
  fetching: StrictDict(fetching),
  gotGradesFrozen,
});
