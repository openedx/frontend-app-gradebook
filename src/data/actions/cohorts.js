import { StrictDict } from '@src/utils';
import { createActionFactory } from './utils';

export const dataKey = 'cohorts';
const createAction = createActionFactory(dataKey);

const fetching = {
  started: createAction('fetching/started'),
  error: createAction('fetching/error'),
  /**
   * fetching.received(results)
   * @param {object[]} results - cohorts fetch results
   */
  received: createAction('fetching/received'),
};

export default StrictDict({
  fetching: StrictDict(fetching),
});
