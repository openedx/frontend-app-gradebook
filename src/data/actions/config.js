import { StrictDict } from '@src/utils';
import { createActionFactory } from './utils';

export const dataKey = 'config';
const createAction = createActionFactory(dataKey);

/**
 * gotBulkManagemmentConfig(bulkManagementAvailable)
 * @param {bool} bulkManagementAvailable - is bulk management available?
 */
const gotBulkManagementConfig = createAction('gotBulkManagement');

export default StrictDict({
  gotBulkManagementConfig,
});
