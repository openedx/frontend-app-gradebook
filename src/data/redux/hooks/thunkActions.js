import { StrictDict } from 'utils';
import thunkActions from 'data/thunkActions';
import { actionHook } from './utils';

const app = StrictDict({
  filterMenu: {
    useCloseMenu: actionHook(thunkActions.app.filterMenu.close),
    useHandleTransitionEnd: actionHook(thunkActions.app.filterMenu.handleTransitionEnd),
    useToggleMenu: actionHook(thunkActions.app.filterMenu.toggle),
  },
  useSetModalStateFromTable: actionHook(thunkActions.app.setModalStateFromTable),
});

const grades = StrictDict({
  useFetchGradesIfAssignmentGradeFiltersSet: actionHook(
    thunkActions.grades.fetchGradesIfAssignmentGradeFiltersSet,
  ),
  useFetchPrevNextGrades: actionHook(thunkActions.grades.fetchPrevNextGrades),
  useFetchGrades: actionHook(thunkActions.grades.fetchGrades),
  useSubmitImportGradesButtonData: actionHook(thunkActions.grades.submitImportGradesButtonData),
  useUpdateGrades: actionHook(thunkActions.grades.updateGrades),
});

export default StrictDict({
  app,
  grades,
});
