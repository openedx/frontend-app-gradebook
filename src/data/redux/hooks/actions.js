import { StrictDict } from 'utils';
import actions from 'data/actions';
import { actionHook } from './utils';

const app = StrictDict({
  useSetLocalFilter: actionHook(actions.app.setLocalFilter),
  useSetSearchValue: actionHook(actions.app.setSearchValue),
  useSetShowImportSuccessToast: actionHook(actions.app.setShowImportSuccessToast),
  useSetView: actionHook(actions.app.setView),
  useCloseModal: actionHook(actions.app.closeModal),
  useSetModalState: actionHook(actions.app.setModalState),
});

const filters = StrictDict({
  useUpdateAssignment: actionHook(actions.filters.update.assignment),
  useUpdateAssignmentLimits: actionHook(actions.filters.update.assignmentLimits),
  useUpdateAssignmentType: actionHook(actions.filters.update.assignmentType),
  useUpdateCohort: actionHook(actions.filters.update.cohort),
  useUpdateCourseGradeLimits: actionHook(actions.filters.update.courseGradeLimits),
  useUpdateIncludeCourseRoleMembers: actionHook(actions.filters.update.includeCourseRoleMembers),
  useUpdateTrack: actionHook(actions.filters.update.track),
  useResetFilters: actionHook(actions.filters.reset),
});

const grades = StrictDict({
  useDoneViewingAssignment: actionHook(actions.grades.doneViewingAssignment),
  useDownloadBulkGradesReport: actionHook(actions.grades.downloadReport.bulkGrades),
  useDownloadInterventionReport: actionHook(actions.grades.downloadReport.intervention),
  useToggleGradeFormat: actionHook(actions.grades.toggleGradeFormat),
  useCloseBanner: actionHook(actions.grades.banner.close),
});

export default StrictDict({
  app,
  filters,
  grades,
});
