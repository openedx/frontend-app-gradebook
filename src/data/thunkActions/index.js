import { StrictDict } from '../../utils';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import grades from './grades';
import roles from './roles';
import tracks from './tracks';

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  grades,
  roles,
  tracks,
});
