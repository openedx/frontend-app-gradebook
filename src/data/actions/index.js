import { StrictDict } from '../../utils';

import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import config from './config';
import filters from './filters';
import grades from './grades';
import roles from './roles';
import tracks from './tracks';

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  config,
  filters,
  grades,
  roles,
  tracks,
});
