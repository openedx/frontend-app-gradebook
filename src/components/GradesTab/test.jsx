import React from 'react';
import { shallow } from 'enzyme';

import actions from 'data/actions';
import thunkActions from 'data/thunkActions';

import {
  GradesTab,
  mapStateToProps,
  mapDispatchToProps,
} from '.';

jest.mock('data/actions', () => ({
  __esModule: true,
  default: {
    filters: { resetFilters: jest.fn() },
  },
}));
jest.mock('data/thunkActions', () => ({
  __esModule: true,
  default: {
    grades: { fetchGrades: jest.fn() },
  },
}));

jest.mock('./PageButtons', () => 'PageButtons');
jest.mock('./FilterBadges', () => 'FilterBadges');
jest.mock('./BulkManagementControls', () => 'BulkManagementControls');
jest.mock('./EditModal', () => 'EditModal');
jest.mock('./GradebookTable', () => 'GradebookTable');
jest.mock('./SearchControls', () => 'SearchControls');
jest.mock('./StatusAlerts', () => 'StatusAlerts');
jest.mock('./SpinnerIcon', () => 'SpinnerIcon');
jest.mock('./ScoreViewInput', () => 'ScoreViewInput');
jest.mock('./UsersLabel', () => 'UsersLabel');

describe('GradesTab', () => {
  let props;
  beforeEach(() => {
    props = {
      updateQueryParams: jest.fn(),
      fetchGrades: jest.fn(),
      resetFilters: jest.fn(),
    };
  });

  describe('Component', () => {
    const filterNames = ['duck', 'Duck', 'Duuuuuck', 'GOOOOSE!'];
    describe('behavior', () => {
      let el;
      beforeEach(() => {
        el = shallow(<GradesTab {...props} />);
      });
      describe('handleFilterBadgeClose', () => {
        beforeEach(() => {
          el.instance().handleFilterBadgeClose(filterNames)();
        });
        it('calls props.resetFilters with the filters', () => {
          expect(props.resetFilters).toHaveBeenCalledWith(filterNames);
        });
        it('calls props.updateQueryParams with a reset-filters obj', () => {
          expect(props.updateQueryParams).toHaveBeenCalledWith({
            [filterNames[0]]: false,
            [filterNames[1]]: false,
            [filterNames[2]]: false,
            [filterNames[3]]: false,
          });
        });
        it('calls fetchGrades', () => {
          expect(props.fetchGrades).toHaveBeenCalledWith();
        });
      });
    });
    describe('snapshots', () => {
      test('basic snapshot', () => {
        const el = shallow(<GradesTab {...props} />);
        el.instance().handleFilterBadgeClose = jest.fn().mockName('this.handleFilterBadgeClose');
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
  });
  test('mapStateToProps is empty', () => {
    expect(mapStateToProps({ some: 'state' })).toEqual({});
  });
  describe('mapDispatchToProps', () => {
    describe('fetchGrades', () => {
      test('from thunkActions.grades.fetchGrades', () => {
        expect(mapDispatchToProps.fetchGrades).toEqual(
          thunkActions.grades.fetchGrades,
        );
      });
    });
    describe('resetFilters', () => {
      test('from actions.filters.reset', () => {
        expect(mapDispatchToProps.resetFilters).toEqual(
          actions.filters.reset,
        );
      });
    });
  });
});
