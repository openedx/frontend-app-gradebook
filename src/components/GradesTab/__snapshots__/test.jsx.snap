// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradesTab Component snapshots basic snapshot 1`] = `
<React.Fragment>
  <SpinnerIcon />
  <SearchControls />
  <FilterBadges
    handleClose={[MockFunction this.handleFilterBadgeClose]}
  />
  <StatusAlerts />
  <h4>
    <FormattedMessage
      defaultMessage="Step 2: View or Modify Individual Grades"
      description="Alert text for invalid minimum course grade"
      id="gradebook.GradesTab.gradebookStepHeading"
    />
  </h4>
  <UsersLabel />
  <div
    className="d-flex justify-content-between align-items-center mb-2"
  >
    <ScoreViewInput />
    <BulkManagementControls />
  </div>
  <GradebookTable />
  <PageButtons />
  <p>
    * 
    <FormattedMessage
      defaultMessage="available for learners in the Master's track only"
      description="Masters feature availability hint on Grades Tab"
      id="gradebook.GradesTab.mastersHint"
    />
  </p>
  <EditModal />
</React.Fragment>
`;
